Mage/src/main/java/mage/util/CardUtil.java

```diff
@@ -1148,6 +1155,116 @@ public static void makeCardPlayable(Game game, Ability source, Card card, Durati
         }
     }
 
+    private static List<SpellAbility> getCastableWithAttributes(Card cardToCast, FilterCard filter, UUID sourceId, UUID playerId, Game game) {
+        List<SpellAbility> spellAbilities = new ArrayList<>();
+        if (cardToCast instanceof SplitCard) {
+            if (filter.match(((SplitCard) cardToCast).getLeftHalfCard(), sourceId, playerId, game)) {
+                spellAbilities.add(((SplitCard) cardToCast).getLeftHalfCard().getSpellAbility());
+            }
+            if (filter.match(((SplitCard) cardToCast).getRightHalfCard(), sourceId, playerId, game)) {
+                spellAbilities.add(((SplitCard) cardToCast).getRightHalfCard().getSpellAbility());
+            }
+        } else if (cardToCast instanceof AdventureCard) {
+            if (filter.match(cardToCast, sourceId, playerId, game)) {
+                spellAbilities.add(cardToCast.getSpellAbility());
+            }
+            if (filter.match(((AdventureCard) cardToCast).getSpellCard(), sourceId, playerId, game)) {
+                spellAbilities.add(((AdventureCard) cardToCast).getSpellCard().getSpellAbility());
+            }
+        } else if (cardToCast instanceof ModalDoubleFacesCard) {
+            if (filter.match(((ModalDoubleFacesCard) cardToCast).getLeftHalfCard(), sourceId, playerId, game)) {
+                spellAbilities.add(((ModalDoubleFacesCard) cardToCast).getLeftHalfCard().getSpellAbility());
+            }
+            if (filter.match(((ModalDoubleFacesCard) cardToCast).getRightHalfCard(), sourceId, playerId, game)) {
+                spellAbilities.add(((ModalDoubleFacesCard) cardToCast).getRightHalfCard().getSpellAbility());
+            }
+        } else {
+            spellAbilities.add(cardToCast.getSpellAbility());
+        }
+        spellAbilities.removeIf(Objects::isNull);
+        return spellAbilities;
+    }
+
+    private static final FilterCard defaultFilter = new FilterCard("card to cast");
+
+    private static TargetCard getTarget(Set<UUID> cardIds, Game game) {
+        Zone zone = game.getState().getZone(cardIds.stream().findFirst().orElse(null));
+        switch (zone) {
+            case HAND:
+                return new TargetCardInHand(0, 1, defaultFilter);
+            case GRAVEYARD:
+                return new TargetCardInGraveyard(0, 1, defaultFilter);
+            case EXILED:
+                return new TargetCardInExile(0, 1, defaultFilter, null);
+            case LIBRARY:
+                return new TargetCardInLibrary(0, 1, defaultFilter);
+            default:
+                throw new UnsupportedOperationException("can't choose a card from zone: " + zone);
+        }
+    }
+
+    public static boolean castSpellWithAttributesForFree(Player player, Ability source, Game game, Cards cards, FilterCard filter) {
```

Comment:
I can't see anti-freeze code for AI. Any anti-infinite cycle code (any `while`, etc) -- AI player must able to quit it on "freeze", e.g. on try to choose wrong choice every time. That's why it's important to remove card otherwise AI will choose a wrong card forever (see #5023 for dozen examples and fixes). Your fixes: remove card from the list on false return (if it hasn't been cast after choose).

![shot_210819_095255](https://user-images.githubusercontent.com/8344157/130015227-5adc5ec9-b07a-461f-9a24-258bedb98007.png)

Reply:
Would it be good enough to use `Outcome.AIDontUseIt` or should there be something else?

Reply:
Nope, AI must able to play it (it's a free cast without limit, so AI can play it without problems and it do in old code). I can modify it later if you want (add card remove on fail cast).

Reply:
so something like this?

![image](https://user-images.githubusercontent.com/29759218/130061423-0e093f14-2754-4d3a-9c6b-4358475764ea.png)

Reply:
Yes, each new cast must remove failed card from the list for AI, so he can't select it again.

P.S. There are also a special `player.isComputer()` call if you need AI specific and test compatible code (example: card logic allows to de-select card from the list before continue, but AI don't need it and must process next).

Reply:
yeah I’ll add an isComputer call in there